### PR TITLE
Implement `SampleConsensusModelSphere<PointT>::projectPoints` properly

### PR DESCRIPTION
**Pull request by fidergo-stephane-gourichon**
Took example from sac_model_circle3d.hpp.
The sphere case is actually simpler than the circle3d one.
Tested here, works. Just rebased for pull request.

### CODE REVIEW

**Comment by taketwo** [comment]
Thanks for contributing. Looking at your code, the difference that `copy_data_fields` argument makes is whether all points are projected, or only inliers. I don't think this is a correct interpretation. I think only the inliers should be projected. And the argument controls whether the the data fields (other than `x`, `y`, `z`) are copied into the new point cloud.

**Comment by fidergo-stephane-gourichon** [comment]
> Thanks for contributing. Looking at your code, the difference that copy_data_fields argument makes is whether all points are projected, or only inliers. I don't think this is a correct interpretation. I think only the inliers should be projected. 

What makes you believe that? The PR code does what you think it should do: it projects only inliers in all cases.

> And the argument controls whether the the data fields (other than x, y, z) are copied into the new point cloud.

The documentation for the method is not very clear IMHO. Here's a snippet:

https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/sac_model.h#L273-L285

The interpretation that is implemented in all existing classes is: if true, copy all points to output and project inliers, else only project inliers to output.

You can check existing implementations for yourself. Here are links in today's master code to all the places that implement `projectPoints`:

* [sac_model_circle3d.hpp line 291](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle3d.hpp#L291)
* [sac_model_circle.hpp line 229](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp#L229)
* [sac_model_cone.hpp line 346](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_cone.hpp#L346)
* [sac_model_cylinder.hpp line 303](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_cylinder.hpp#L303)
* [sac_model_line.hpp line 234](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_line.hpp#L234)
* [sac_model_plane.hpp line 261](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp#L261)
* [sac_model_stick.hpp line 264](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_stick.hpp#L264)

This PR strictly follows the existing pattern. You can do a visual diff to check for it. The structure is the same. **Actually, I wrote the new implementation as a merge with manual adjust from [sac_model_circle3d.hpp line 291](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle3d.hpp#L291)**.

In pre-PR code all classes do the same, except `sac_model_sphere.hpp` which only outputs a warning.
There might be something unclear in the way it's currently implemented with respect to the documentation, but at least this PR increases consistency: post-PR, all classes implement `projectPoints()` consistently.

Does this address the two issues you raised? If anything remains after this PR, is really a different issue common to all implementations, to be dealt with separately.

Thank you.

**Review by taketwo — DISMISSED** [review]

**Comment by SergioRAgostinho** [comment]
Are there unit tests validating the newly implemented behavior? This [file](https://github.com/PointCloudLibrary/pcl/blob/master/test/sample_consensus/test_sample_consensus_quadric_models.cpp) would be likeliest place for them to exist, but I don't see any call to `projectPoints` in there.

Basically I just want to make sure there are unit tests validating the new functionality @fidergo-stephane-gourichon just added.

**Comment by fidergo-stephane-gourichon** [comment]
@SergioRAgostinho as far as I can tell, implementations of projectPoints are tested for those classes:

SampleConsensusModelPlane
SampleConsensusModelNormalPlane
SampleConsensusModelLine
SampleConsensusModelParallelLine

They are not currently tested for those classes:

SampleConsensusModelNormalParallelPlane
SampleConsensusModelSphere <- the one I implemented
SampleConsensusModelNormalSphere
SampleConsensusModelCone
SampleConsensusModelCylinder
SampleConsensusModelCircle2D
SampleConsensusModelCircle3D

**Comment by SergioRAgostinho** [comment]
Can you please add some?

Generate a trivial synthetic test case, easy to understand:
1. create a simple cloud manually
2. set the model coefficients manually.
3. run the projection method and check the results.

**Comment by mvieth** [comment]
@fidergo-stephane-gourichon Are you interested in completing this pull request? It seems like mainly the test case is missing

**Comment by fidergo-stephane-gourichon** [comment]
Hi @mvieth. Thanks for asking. I am no longer involved with the project that caused me to contribute this code, and am now busy with other things. So I cannot contribute with tests in the near future. I understand that test coverage is a desirable property, pushing for merged code to come with tests as often as possible. Still, I think the code itself is fine (it worked for us IIRC) and, if still relevant to the current code base I'd be happy for it to be merged.

**Comment by mvieth** [comment]
Alright, that's understandable. Then I might implement the test sometime.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:427]:
```suggestion
    for (std::size_t i = 0; i < inliers.size (); ++i)
```

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:432]:
```suggestion
    for (std::size_t i = 0; i < inliers.size (); ++i)
```

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:442]: Same comment as aboove

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:404]: Declared outside the if-else blocks

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:438]: Declared outside the if-else block

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:436]: vector3fMap?

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:402]: vector3fMap?

**Comment by mvieth** [comment]
@kunaltyagi How should be proceed here, since the original creator of this PR can not/does not want to work further on this? Either we make the changes directly on the existing branch, or should be create a new pull request?

**Comment by kunaltyagi** [comment]
We should be able to take over this PR as maintainers. A new PR is ok as well. Usually continuing is just easier

File: test/sample_consensus/test_sample_consensus_quadric_models.cpp
Comment (kunaltyagi) [inline test/sample_consensus/test_sample_consensus_quadric_models.cpp:695]: I thought it would be simpler to just compare if a point is inlier or not, but it turns out, it is not. Take it or leave it 😆 

```suggestion
  for (int i = 0, j = 0; i < projected_points_all.size(); ++i) {
    if (j < inliers.size() && i == inliers[j]) {  // idx is an inlier
      EXPECT_XYZ_NEAR(projected_points_all[i], projected_truth[j], 1e-5);
      ++j;  // check for next inlier
    } else {  // not an inlier, so not projected
      EXPECT_XYZ_NEAR(projected_points_all[i],        (*input)[i], 1e-5);
    }
  }
```

**Comment by kunaltyagi** [comment]
Test looks good :) 

Wrt labels, this is a fix for a missing feature, right?

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:435]: Not easily usable because the rest is in double precision. A cast would be possible of course, but this line is is not so bad IMO.
I applied most of the other changes you suggested, feel free to make further improvements

File: test/sample_consensus/test_sample_consensus_quadric_models.cpp
Comment (mvieth) [reply test/sample_consensus/test_sample_consensus_quadric_models.cpp:695]: IMO neither shorter nor more understandable, so I see no advantage TBH

**Review by kunaltyagi — DISMISSED** [review]
LGTM if CI is green

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (larshg) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:405]: Would it give some, to sprinkle const infront of these 3 eigens?

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (larshg) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp:435]: Same here?

**Review by mvieth — APPROVED** [review]

**Comment by fidergo-stephane-gourichon** [comment]
Hi! I browsed your conversations and changes. Wow, that's maintainer activity! Happy to see my code nearly merged. :-)

**Review by larshg — APPROVED** [review]

**Pull request merged**
